### ISSUE TITLE
Can't access requester pays buckets

### ISSUE THREAD

**Issue by rsignell-usgs** [Quansight/qhub#162]
Accessing "requester pays" S3 buckets is used in most of my workflows, but I can't get them to work here on my qhub deployment on AWS.   

For example, this works for me on other pangeo deployments on AWS:
```python
import xarray as xr
import fsspec

ds = xr.open_zarr(fsspec.get_mapper('s3://pangeo-data-uswest2/esip/adcirc/ike',
            anon=False, requester_pays=True))
```
But when I try this on qhub, I get: 
```python-traceback
---------------------------------------------------------------------------
ClientError                               Traceback (most recent call last)
/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/s3fs/core.py in _call_s3(self, method, *akwarglist, **kwargs)
    206             try:
--> 207                 return await method(**additional_kwargs)
    208             except S3_RETRYABLE_ERRORS as e:

/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/aiobotocore/client.py in _make_api_call(self, operation_name, api_params)
    150             error_class = self.exceptions.from_code(error_code)
--> 151             raise error_class(parsed_response, operation_name)
    152         else:

ClientError: An error occurred (AccessDenied) when calling the GetObject operation: Access Denied

The above exception was the direct cause of the following exception:

PermissionError                           Traceback (most recent call last)
<ipython-input-2-57eee302fb24> in <module>
      1 ds = xr.open_zarr(fsspec.get_mapper('s3://pangeo-data-uswest2/esip/adcirc/ike',
----> 2             anon=False, requester_pays=True))

/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/xarray/backends/zarr.py in open_zarr(store, group, synchronizer, chunks, decode_cf, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, consolidated, overwrite_encoded_chunks, chunk_store, decode_timedelta, use_cftime, **kwargs)
    665         group=group,
    666         consolidated=consolidated,
--> 667         chunk_store=chunk_store,
    668     )
    669     ds = maybe_decode_store(zarr_store)

/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/xarray/backends/zarr.py in open_group(cls, store, mode, synchronizer, group, consolidated, consolidate_on_close, chunk_store)
    290             zarr_group = zarr.open_consolidated(store, **open_kwargs)
    291         else:
--> 292             zarr_group = zarr.open_group(store, **open_kwargs)
    293         return cls(zarr_group, consolidate_on_close)
    294 

/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/zarr/hierarchy.py in open_group(store, mode, cache_attrs, synchronizer, path, chunk_store, storage_options)
   1183 
   1184     return Group(store, read_only=read_only, cache_attrs=cache_attrs,
-> 1185                  synchronizer=synchronizer, path=path, chunk_store=chunk_store)

/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/zarr/hierarchy.py in __init__(self, store, path, read_only, chunk_store, cache_attrs, synchronizer)
    114         try:
    115             mkey = self._key_prefix + group_meta_key
--> 116             meta_bytes = store[mkey]
    117         except KeyError:
    118             raise GroupNotFoundError(path)

/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/fsspec/mapping.py in __getitem__(self, key, default)
    130         k = self._key_to_str(key)
    131         try:
--> 132             result = self.fs.cat(k)
    133         except self.missing_exceptions:
    134             if default is not None:

/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/fsspec/asyn.py in cat(self, path, recursive, on_error, **kwargs)
    228             ex = next(filter(is_exception, out), False)
    229             if ex:
--> 230                 raise ex
    231         if (
    232             len(paths) > 1

/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/s3fs/core.py in _cat_file(self, path, version_id)
    596         resp = await self._call_s3(
    597             self.s3.get_object, Bucket=bucket, Key=key,
--> 598             **version_id_kw(version_id or vers),
    599         )
    600         data = await resp['Body'].read()

/home/conda/store/66bc3279f742db6a733b4f5a5b8d4897a663a5ebcf016590a700adabb81a81fb-pangeo/lib/python3.7/site-packages/s3fs/core.py in _call_s3(self, method, *akwarglist, **kwargs)
    223             except Exception as e:
    224                 err = e
--> 225         raise translate_boto_error(err) from err
    226 
    227     call_s3 = sync_wrapper(_call_s3)

PermissionError: Access Denied
```

I noticed that I didn't seem to have any AWS credentials, so I set them with `aws configure`, but that still didn't help! 

Any ideas @aktech or @prasunanand ?

**Comment by costrouc** [comment]
@rsignell-usgs it I am understanding this correctly I this is an issue outside of the scope of qhub. When you launch a jupyterlab instance it doesn't not (at least in default configuration) pass any aws credentials. You will need to add the aws credentials to your home directory. 

`~/.aws/credentials`

```ini
[default]
aws_access_key_id = ....
aws_secret_access_key = ....
```

Assuming these credentials have sufficient permissions you should be able to access the s3 bucket. You can also do it via environment variables https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html

**Comment by rsignell-usgs** [comment]
@costrouc, I had set the `~/.aws/credentials` via `aws configure` but as you mention, the "access denied" problem turned out to be an issue with the latest version of `s3fs`.   When I reverted to `s3fs=0.4.2`, I was able to open the zarr dataset in the requester pays bucket from the jupyter notebook.  

When I try to use the dask cluster, however, the read fails.  Is it possible the Dask workers don't access the AWS credentials?

**Comment by rsignell-usgs** [comment]
When I look at our other non-qhub Pangeo deployments on AWS, I see these environment variables defined on login:
```
(notebook) jovyan@jupyter-rsignell-2dusgs:~$ env | grep AWS
AWS_WEB_IDENTITY_TOKEN_FILE=/var/run/secrets/eks.amazonaws.com/serviceaccount/token
AWS_ROLE_ARN=arn:aws:iam::783380859522:role/eksctl-pangeo-addon-iamserviceaccount-icesat-Role1-1BKE2MPBN625A
```
and the role includes this policy to read/list any accessible bucket and write to some specific buckets:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "ReadAnyWriteSome",
            "Effect": "Allow",
            "Action": [
                "s3:List*",
                "s3:Get*"
            ],
            "Resource": [
                "*"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::pangeo-data-upload-oregon/*",
                "arn:aws:s3:::pangeo-scratch/*"
            ]
        }
    ]
}
```
This allows the Dask Gateway workers to read from requester pays buckets as well as write to those specific buckets, and without the need to create `~/.aws/credentials` file.  

I'm sure you know this, but I'm including this link to the AWS announcement of [fine-grained IAM roles for service accounts](https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/) and the [Terraform aws_iam_openid_connect_provider](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_openid_connect_provider). 

Seems like this kind of access to S3 would be useful for Qhub users as well?

**Comment by rsignell-usgs** [comment]
Woohoo, I figured it out.    @jkellndorfer tipped me off that there were EKS roles already, and I just opened up
arn:aws:iam::313654087598:role/aws-service-role/eks-nodegroup.amazonaws.com/AWSServiceRoleForAmazonEKSNodegroup and added the S3 policy defined above.  

And now that they have the added policy added to their role, my Dask workers can read from requester_pays buckets!

![2020-11-09_12-43-26](https://user-images.githubusercontent.com/1872600/98577097-75f84180-2289-11eb-9dba-3c0c443233e6.png)
